### ISSUE TITLE
Adding time derivatives of some metrics

### ISSUE THREAD

**Issue by sciaba** [HSF/prmon#64]
Hi,
I would humbly suggest to add time derivatives of some metrics, namely the IO and the network metrics. As we know, spikes in IO rates are extremely important when characterising an application and having prmon_plot.py being able to generate plots for them would be very handy.
IMO it can be done just in prmon_plot.py, no need to touch prmon.
Cheers,
Andrea

**Comment by amete** [comment]
Hi Andrea,

If I understand your request correctly the core functionality should be there already, i.e.:

```
prmon_plot.py --input ... --xvar wtime --yvar rx_bytes
```

would create the usual time evolution of `rx_bytes` like:

![prmon_wtime_vs_rx_bytes](https://user-images.githubusercontent.com/7428938/41039829-edef094e-699a-11e8-84d1-f9818cea5289.png)

while

```
prmon_plot.py --input ... --xvar wtime --yvar rx_bytes --diff
```

would compute and plot delta(yvar)/delta(xvar) as:

![prmon_wtime_vs_diff_rx_bytes](https://user-images.githubusercontent.com/7428938/41039890-1de7b952-699b-11e8-9b40-4d2783b8379a.png)

I think we can improve the labeling etc. but is this what you're after?

Best,
Serhan

**Comment by sciaba** [comment]
Exactly! The only degree of freedom is the delta_t used to estimate the derivative. It should be a multiple of the sampling interval; some degree of smoothing may be desirable (maybe a sensible default and a configurable parameter? Up to you!).
Cheers,
Andrea

**Comment by sciaba** [comment]
I was thinking again, and maybe it's not just a matter of plots, but we should consider IO rates as "real" metrics, which means having their values written already in prmon.txt, such that prmon.json will contain max and avg of them, too... Clearly it's more work because they should be calculated in prmon itself.
Andrea

**Comment by graeme-a-stewart** [comment]
Hi @sciaba - I am not so sure about that. From the accumulating values it's easy to calculate the derivative and we could easily adopt a few strategies to do that (various sliding window shapes and widths). If you take the derivative inside prmon and then only give rates it's not so easy to go backwards.

Technically it's not *that* hard to do, but I what we do now is more useful and reflects the underlying metric better, so it's easier to relate to other measurements.

**Comment by sciaba** [comment]
You convinced me!

**Comment by amete** [comment]
The plotting aspect of this issue is addressed in #66. I think the consensus is not to make any changes to the core code for this. Therefore, I'm closing this issue. Please feel free to re-open if need be.
